Release v3.6.0

AlvinSchiller opened the pull request:
### Features
- #1943 Event Device Support (evdev)
- #1970 Shutdown after [Time] idle
  - #2332 Add Idle Shutdown Timer support
- #2352 Allow to enable/disable Cover Art in Web App & load from filesystem
- #2425 lift 64bit restriction
- #2557 Improve IRQ configuration and fallback handling for RC522 SPI RFID reader
- #2470 Replace RPi.GPIO with rpi-lgpio
  - #2568 fix lgpio build for armv6 and python 3.13
- #2412 add basic mqtt support
- #2380 Add INA219 battery sensor

### Fixes
- #2239 github actions update
- #2284 Add markdown linting
- #2292 search filter in library
- #2293 Stop Back Action at audiofolder root level
- #2307 Build libzmq locally to work on all host types
- #2320 Add Python 3.12 to Action
- #2323 Fix PlayerMPD.rewind to start with the first song
- #2325 Fix CoverartCacheManager
- #2326 Fix PlayerMPD.prev/next() when stopped
- #2351 docker build after pyzmq update
- #2498 fix evdev breakingchange
- #2444 RDM6300 configuration bug 
  - #2450 Use correct raspi-config command for bookworm
- #2339  fix bad utils.get_config_action invocations

### Documentation
- #2257 Extract docs for battery monitor
- #2284 Add markdown linting
- #2300 Manually upgrade to the latest version
- #2301 Fix GPIO shutdown button naming
- #2484 Fix the url to Autohotspot docs
- #2616 installation instructions for new imager version

AlvinSchiller committed:
Bump to v3.6.0-alpha

s-martin committed:
Extract docs for battery monitor (#2257)

* extract battmon docs into markdown

* Fix typo

* add link to ADS1015

AlvinSchiller committed:
github actions update (#2239)

* Bump actions/download-artifact from 3 to 4

Bumps [actions/download-artifact](https://github.com/actions/download-artifact) from 3 to 4.
- [Release notes](https://github.com/actions/download-artifact/releases)
- [Commits](https://github.com/actions/download-artifact/compare/v3...v4)

* Bump actions/upload-artifact from 3 to 4

Bumps [actions/upload-artifact](https://github.com/actions/upload-artifact) from 3 to 4.
- [Release notes](https://github.com/actions/upload-artifact/releases)
- [Commits](actions/upload-artifact@v3...v4)

* Bump geekyeggo/delete-artifact from 2 to 4

* add write permission for artifact deletion

* ignore fails for artifact deletion

* Revert "ignore fails for artifact deletion"

This reverts commit a759d89fc0e9e091e0bd11ff7c94cff96c0a5ff6.

* Revert "add write permission for artifact deletion"

This reverts commit 102840204290fb6543fc5d8f8245dca1ac21c63e.

* remove cleanup stage for now

s-martin committed:
(doc) update the docstring markdown file (#2259)

s-martin committed:
(maint) Remove duplicate rst  (#2269)

* remove duplicate rst doc file

* fix link

AlvinSchiller committed:
Merge branch 'future3/develop' into future3/merge-hotfix-3.5.2

AlvinSchiller committed:
Merge pull request #2273 

merge hotfix 3.5.2

votti committed:
Event Device Support (`evdev`) (#1943)

* Add event device plugin

This implements an "event device" listener plugin,
that enables the user to configure the phoniebox to
respond to events from an "event device" (device under /dev/input).
This incluces eg button presses from an USB controller or keyboard.

* Adds documentation for the event device plugin

Includes a detailed how-to as well as example config

* Allow empty button config

This is an actual usecase in case someone wants to setup a device
and figure out the button ids by looking at the logs

* Update README.rst

* Improve README

- Remove duplicated section as suggested
- Add Zero Delay Arcade USB Encoder usecase

* Give more meaningfull name to listener thread variable

* Link to old documentation

* Remove README and fix typo

The documentation is now part of the markdown documentation

* Remove unecessary plain=1

* Remove indents in example code

There was some left over indentation from the old rtf format in the python code examples.

* Add example how to access evdev in docker

* Example for new Evdev config structure

This is tighly modeled after gpio.
In contrast to GPIO, this is a list of devices containing
each one or more input/output devices.
Eg a Joystick has keys but potentially also rumble or leds.

Currently only very simply input devices (buttons) are supported.
This structer should enable to extend this easily.

* Move evdev config to separate file

As suggested this should be a separate file, similar to the gpioz config

* Adapt evdev code to new config

Now the config for evdev is akin to GPIO.
The big difference is that multiple devices with each
multiple input/output devices are supported.

Note that the backend is still the old event device listener backend.
Thus to support more features, the backend would need to be
quite drastically overhauled.

* Update documentation for new evdev config format

* Fix wrong reference

* Fix bug to correctly check supported input device type

* Add more tests for evdev init

* Fix wrong attribute name

The attribute 'device_request' of EvDevKeyListener is now called 'device_name_request'

* Validate input config better

Fails if there is no device_name and setts proper default for 'exact'

* Add pyzmq installation for github action

This is required for some tests.

Following the instructions from issue #2050
Specifically: https://github.com/MiczFlor/RPi-Jukebox-RFID/issues/2050#issuecomment-1804785695

* Revert "Add pyzmq installation for github action"

This reverts commit 75161b545a7a127a5d486c0770c449fbbd52c3d4.

As this was not working

* Mock `jukebox.publishing` in tests to avoid zmq

Currently it is hard to install zmq (#2050) and for
CI pytest `zmq` is not available.
As the test do not really require `jukebox.publishing`
running, mocking it in the test avoids it being imported.

Thus the tests do not require `zmq` to be installed.

---------

Co-authored-by: pabera <1260686+pabera@users.noreply.github.com>
Co-authored-by: Vito Zanotelli <vito.zanotelli@gmail.com>
Co-authored-by: s-martin <s-martin@users.noreply.github.com>

pabera committed:
Fix search filter in library #2290 (#2292)

pabera committed:
fix: Stop Back Action at audiofolder root level (#2293)

* bugfix: Stop Back Action at audiofolder root level

This also allows to go back when folder is empty

Fixes #2224

* fix: Flake8 issue

* Remove some minor issues

AlvinSchiller committed:
Reactivate delete artifact  (#2297)

* activated delete-artifacts again with v5

* deactivate failOnError for deletion

hoffie committed:
(docs) fix gpio shutdown button naming (#2301)

s-martin committed:
Add markdown linting (#2284)

* fix docs

* add markdown action

* use different action

* change linter

* add checkout

* change globs

* change wildcard

* Aktualisieren von markdown_v3.yml

* fix ignore

* comment ignore

* change version

* use master

* new action used for linting

* Fix config

* Remove ignore

* Ignore docstring dir

* ignore GitHub

* Aktualisieren von markdown_v3.yml

* use dot true

* Aktualisieren von markdown_v3.yml

* Use new action

* Fix continue-on-error

* Remove unnecessary code

* Add markdownlint config

* Rename . markdownlint-cli2.yaml to .markdownlint-cli2.yaml

* Use config file

* Simplify globs

* Use globs and ignore in config file

* Ignore $ errors

* Ignore src and .github

* Use only config file

* Fix general glob

* Ignore inline html

* Disable MD010

* Fix language

* dont warn trailing punctuation

* fix warnings

* fix a warning

* reduce ignore

* fix warnings

* fail check, if markdownlint fails

* add runner script and hook

* Update documentation/builders/components/power/onoff-shim.md

Co-authored-by: Alvin Schiller <103769832+AlvinSchiller@users.noreply.github.com>

* Update playlists-livestreams-podcasts.md

* add documentation for documentation with md

* get rid of docker

* fix comments

* incorporate comments

* Update documentation/developers/documentation.md

Co-authored-by: Alvin Schiller <103769832+AlvinSchiller@users.noreply.github.com>

* check, if cmd does not exisr

---------

Co-authored-by: Alvin Schiller <103769832+AlvinSchiller@users.noreply.github.com>

pabera committed:
docs: Manually upgrade to the latest version (#2300)

* docs: Manually upgrade to the latest version

* Fix typo

* Fix Typos

* Solve last issue

* fix: Align with comments

* fix: indentation was off

* fix: markdown lint

* fix: one small bugfix

* fix: remove last linting issues

* fix: adding info about missing upgrade feature

s-martin committed:
Customize markdownlint (#2310)

* Customize markdownlint

* Remove ignore in file

* Remove ignore in file

s-martin committed:
Update CodeQL to v3 (#2312)

* use codeql v3

* make sure there's always an update

pabera committed:
fix: Build libzmq locally to work on all host types (#2307)

* Externalize libzmq build

* Update documentation

* Use dedicated build directory for libzmq

* improve: update make command to be more efficient

* Update docker/Dockerfile.libzmq

Co-authored-by: notapirate <notapirate@users.noreply.github.com>

* fix: Remove unneccesary MD linting rule

* fix: Remove another markdown linter

* Remove uncessary file

* refactor: Rename docker.pulse.mpd.conf

---------

Co-authored-by: notapirate <notapirate@users.noreply.github.com>

hoffie committed:
Fix PlayerMPD.rewind to start with the first song (#2323)

rewind() is documented to jump to the first song of the playlist.
Instead, it jumped to the second song as SONGPOS in
MPDClient.play(SONGPOS) is zero-indexed [1], so mpd.play(1) started
the second song.

[1] https://mpd.readthedocs.io/en/latest/protocol.html#the-queue
"The position is a 0-based index"

Related: #2294

s-martin committed:
Add Python 3.12 to Action (#2320)

* Try python 3.12

* Trigger py file for action

* Revert previous commit

* fix flake8 warnings for python 3.12

* fix e126

* fix e126

* ignoring e126 as it is too strict

* revert e126 attempts

hoffie committed:
Fix CoverartCacheManager (#2325)

* Fix CoverartCacheManager for songs with no art

Previously, an ERROR was logged for each song without cover art when the
Web UI was open.
This commit avoids the error, caches the no-cover-art result and saves a
roundtrip to mpd for all no-cover-art songs.

* refactor: Reducing code and simplifying some logical statements

* fix: flake8 error

* refactor: reducing complexity for cache filename

* refactor: introduce queuing for saving cache files

* fix: remove slugify

* feat: Use mutagen instead of MPD to retrieve cover art, include cache flush, and thread

* fix: flake8 error

* Update src/jukebox/components/playermpd/__init__.py

Co-authored-by: Christian Hoffmann <christian@hoffie.info>

---------

Co-authored-by: pabera <1260686+pabera@users.noreply.github.com>

s-martin committed:
maint: Update actions (#2334)

* Update action versions

* Trigger python action

* Update Python action

* Use correct env variable

* Revert change for triggering action

hoffie committed:
Fix PlayerMPD.prev/next() when stopped (#2326)

* utils: Add get_config_action

This abstracts away the functionality to resolve a given config option
to an action in a pre-defined dict.

Co-authored-by: Christian Hoffmann <christian@hoffie.info>

* Fix PlayerMPD.prev/next() when stopped

* Avoid MPD-related crashes during all prev/next() calls.
* Explicitly handle prev() in stopped state, configurable via
  `playermpd.stopped_prev_action`.
* Explicitly handle next() in stopped state, configurable via
  `playermpd.stopped_next_action`.
* Explicitly handle next() when reaching the end of the playlist:
  jukebox-daemon will now ignore the action by default (similar to v2).
  It can also be configured to rewind the playlist instead by setting
  the new config option `playermpd.end_of_playlist_next_action: rewind`
  or to stop playing.

Fixes #2294
Fixes #2327

Co-authored-by: pabera <1260686+pabera@users.noreply.github.com>

---------

Co-authored-by: pabera <1260686+pabera@users.noreply.github.com>

s-martin committed:
fix a typo (#2336)

hoffie commented:
@pabera Another thing which is worth mentioning after #2325 is that the new mutagen dependency will be missing by default. This crashes the player thread upon swipe:

```
Apr 14 22:25:23 phoniebox bash[1026]:   File "/home/phoniebox/RPi-Jukebox-RFID/src/jukebox/components/playermpd/coverart_cache_manager.py", line 1, in <module>
Apr 14 22:25:23 phoniebox bash[1026]:     from mutagen.mp3 import MP3
Apr 14 22:25:23 phoniebox bash[1026]: ModuleNotFoundError: No module named 'mutagen'
```

This fixes it and should probably be run after the update, either automatically (is there a process already?) or via documentation:
`./.venv/bin/pip install -r requirements.txt`

pabera commented:
Yes this is true. We should mention this in the release notes.

At the moment, we don't offer an update mechanism. We have thought about it and will try for the next version.

For the time being, we officially support New Install only (where the dependency will be installed).

hoffie committed:
fix: bad utils.get_config_action invocations (#2339)

AlvinSchiller committed:
Merge branch 'future3/develop' into future3/merge-hotfix-3.5.3

AlvinSchiller committed:
Merge pull request #2346 Merge Hotifx 3.5.3

Merge Hotifx 3.5.3

pabera committed:
fix: docker build after pyzmq update (#2351)

* fix: Add cmake as requirement to support pyzmq build

* fix: Remove cmake and pin pyzmq<26 as cmake does not build well in RPI env

pabera committed:
feat: Allow to enable/disable Cover Art in Web App & load from filesystem (#2352)

* feat: Introduce Web App Settings

* feat: Allow to enable/disable Cover Art in Web App

* fix: handle Falsy mimetype and data even when APIC tag has been found my mutagen

* feat: Try to load cover from filesystem when not found in audio file

* fix: flake8 linting error

* docs: Add documentation for Cover Art

* feat: Allow show_covers setting to be managed in Web App

* fix: again flake8 linting errors

s-martin committed:
Remove setup-python-dependencies (#2363)

s-martin committed:
Fix links (#2366)

s-martin committed:
add details about cards for RC522 (#2372)

* add details about cards

* fix empty line

* Update documentation/developers/rfid/mfrc522_spi.md

Co-authored-by: Alvin Schiller <103769832+AlvinSchiller@users.noreply.github.com>

---------

Co-authored-by: Alvin Schiller <103769832+AlvinSchiller@users.noreply.github.com>

s-martin committed:
Remove unused read of parameter (#2382)

saeugetier committed:
feat: Add INA219 battery sensor (#2380)

* specify specific pip version wich works with zmq install

* add INA219 sensor

* average measurement for more accurate results. Use supply voltage for measurement.

* Revert "specify specific pip version wich works with zmq install"

This reverts commit 48dd1bfc6c73f9a6fb1329ae6528e495b07b36e5.

* correct format

* Update src/jukebox/components/battery_monitor/batt_mon_i2c_ina219/__init__.py

Co-authored-by: s-martin <s-martin@users.noreply.github.com>

* Update copyright notice

* Update license

* Update license

* Document the INA219

* add error handling and type safety

* Update batterymonitor.md

* Update __init__.py

* fix markdown lint for batterymonitor.md

* fix markdown lint batterymonitor.md

* Update documentation/builders/components/power/batterymonitor.md

Co-authored-by: s-martin <s-martin@users.noreply.github.com>

* Update documentation/builders/components/power/batterymonitor.md

Co-authored-by: s-martin <s-martin@users.noreply.github.com>

---------

Co-authored-by: Timm <sauegetier@googlemail.com>
Co-authored-by: s-martin <s-martin@users.noreply.github.com>

c0un7-z3r0 committed:
feat: add initial mqtt plugin

c0un7-z3r0 committed:
feat: add commands and support legacy

c0un7-z3r0 committed:
refactor: tidy up the code and improve logs

c0un7-z3r0 committed:
refactor: tidy up the code and improve logs

pabera committed:
Docker semantic fix (#2451)

mittler-works committed:
Future3/lift 64bit restriction (#2425)

* Remove 64bit check in installer script

* Check presence of raspi.list in raspian check

* fix: update check for debian derived os. integrate checks

RPiOS: 32-bit = raspbian, 64-bit = debian

---------

Co-authored-by: Alvin Schiller <103769832+AlvinSchiller@users.noreply.github.com>

s-martin committed:
Use correct raspi-config command for bookworm (#2450)

* use correct raspi-config command for bookworm

* simplify _get_boot_file_path

* fix path

* revert last commit

* merge fixes

* merge fixes

---------

Co-authored-by: Alvin Schiller <103769832+AlvinSchiller@users.noreply.github.com>

hoffie committed:
feat: Add Idle Shutdown Timer support (#2332)

* feat: Add Idle Shutdown Timer support

This adds an optional idle shutdown timer which can be enabled
via timers.idle_shutdown.timeout_sec in the jukebox.yaml config.

The system will shut down after the given number of seconds if no
activity has been detected during that time. Activity is defined as:
- music playing
- active SSH sessions
- changes in configs or audio content.

Fixes: #1970

* refactor: Break down IdleTimer into 2 standard GenericMultiTimerClass and GenericEndlessTimerClass timers

* feat: Introducing new Timer UI, including Idle Shutdown

* refactor: Abstract into functions

* Adding Sleep timer / not functional

* Finalize Volume Fadeout Shutdown timer

* Fix flake8

* Fix more flake8s

* Fix small bugs

* Improve multitimer.py suggested by #2386

* Fix flake8

---------

Co-authored-by: pabera <1260686+pabera@users.noreply.github.com>

s-martin committed:
ignore LICENSE in markdown linter

s-martin committed:
Merge pull request #2455 from s-martin/future3/ignore-file-in-linter

[maint] Ignore LICENSE file in markdown linter

c0un7-z3r0 committed:
chore: remove not needed line

c0un7-z3r0 committed:
feat: allow to disable mqtt via jukebox.yml

c0un7-z3r0 committed:
docs: add mqtt documentation

c0un7-z3r0 committed:
docs: remove trailing space

c0un7-z3r0 committed:
docs: remove trailing space and fix url

pabera committed:
fix: correct function name reference in is_debian_version_at_least() (#2464)

c0un7-z3r0 committed:
docs: add missed mqtt configurations to example

AlvinSchiller committed:
fix: fix wrong usage of variable (#2473)

c0un7-z3r0 committed:
fix: remove hardcoded base topic in cmd subscription

s-martin committed:
Fix the url to Autohotspot docs

s-martin committed:
Merge pull request #2484 from MiczFlor/future3/fix-help-url

Fix the url to Autohotspot docs

s-martin committed:
Tick MQTT Feature

s-martin committed:
Merge pull request #2412 from c0un7-z3r0/future3/add-mqtt

feat: add basic mqtt support

s-martin committed:
Replace RPi.GPIO with rpi-lgpio for V3 future3 (#2470)

* replace RPi.GPIO with rpi-lgpio

* fix: update setup for RPi.GPIO replacement

* fix: update RFIDReader pi-rc522 Lib for gpiozero support. remove RPi.GPIO references

* fix: pin "pi-rc522-gpiozero" lib to specific commit

* fix: remove RPi.GPIO usage. Check for command existance and ignore if not available

* fix: warning on gcc build

"format ‘%ld’ expects argument of type ‘long int’, but argument 2 has type ‘size_t’ {aka ‘unsigned int’}"

* fix: update docs

* fix: update comment

* Revert "fix: warning on gcc build"

This reverts commit 0be5cdb0e1b6f65ecb690cf01b83c492202b22cf.

* fix: remove obsolete python setup.

* fix: update description for excluded libs

* fix: moved lgpio build into function. fixed needed packages. use download function

* fix: enabled output for wget.

---------

Co-authored-by: Alvin Schiller <103769832+AlvinSchiller@users.noreply.github.com>

AlvinSchiller committed:
fix: "device.fn" was removed in evdev 1.9.0. Use "device.path" as adviced (#2498)

astnerdev committed:
Fix legacy mode by switching to polling when IRQ is unavailable

Legacy mode still called wait_for_tag(), causing the reader to block if no IRQ
pin was wired (common on many RC522 boards and Phoniebox setups). This change
ensures polling is used automatically when IRQ is not present, restoring proper
card detection in legacy mode.

astnerdev committed:
rc522: Add polling mode support when IRQ pin is disabled

Allow pin_irq=0 to disable interrupt-driven card detection and fall back
to polling mode. Store pin_irq as instance variable and add logging to
indicate active mode.

astnerdev committed:
Merge branch 'future3/develop' into fix/rc522-legcy

astnerdev committed:
Clarify IRQ usage and improve defaults for RC522

Updated prompts and logs to warn about high CPU usage in polling mode

Set default IRQ pin to 24 instead of 0

Improved messaging for interrupt-driven card detection

astnerdev committed:
fixed E303 too many blank lines (3)

AlvinSchiller committed:
fix flake8 config (#2566)

* fix: ignore new F824 linting error.
should be adderessed later on

* fix: add files to workflow changes detection

* fix: add requirements to change detection

AlvinSchiller committed:
v3 - fix lgpio build for armv6 and python 3.13 (#2568)

* fix: always build lgpio cause of incomplete wheels

* fix: add setuptools dep for lgpio build

* fix: build lgpio from pypi sources

* fix: always build lg locally

* fix: update no-binary definition

* fix: update comments

* feat: add python 3.13 to test matrix

* fix: run workflow on workflow file changes

* test: add lg lib to ci workflow

AlvinSchiller committed:
Merge branch 'future3/develop' into pr/astnerdev/2557

s-martin committed:
Merge pull request #2557 from astnerdev/future3/fix-rc522-legcy

Improve IRQ configuration and fallback handling for RC522 SPI RFID reader

AlvinSchiller commented:
@pabera @s-martin Are there any tickets that needs to be done before this version can be released?
The open tickets in the above list would be moved to the next version.

s-martin commented:
No.
Release this version. It's about time ;)

AlvinSchiller committed:
Update docs (#2616)

* docs: update installation instruction

* docs(#2425): lift 64bit restriction

* fix: linter MD009

* docs: update mfrc522 reader doc

* fix: linter

* fix: remove left overs

* fix: typo

* fix: update lists

* fix: typo

* fix: improve readability

* fix: linter MD049

AlvinSchiller committed:
Update docs (update.md) and githook (#2623)

* docs: update notes for version updates

* fix: linter MD031

* fix: requirement-excluded info

* fix: packages-core info

* fix: update git-hook "post-merge"

* docs: add overview. update header

* docs: update instructions

* docs: update caution info
